File: Haskell/pi_sequential_map_sum.hs
--  Haskell implementation of π by Quadrature using a map and then a reduce (actually sum).
--
--  Copyright © 2015  Russel Winder

module Main where

import Output (out)

f :: Double -> Int -> Double
f delta n = 1.0 / (1.0 + x * x)
  where
    x = ((fromIntegral n) - 0.5) * delta

main :: IO()
main =  out "Sequential Map Sum" pi n
  where
    n = 100000000 -- 10 times fewer than C++ for speed reasons.
    delta = 1.0 / (fromIntegral n)
    pi = 4.0 * delta * sum (map (f delta) [1..n])

File: Rust/src/bin/pi_sequential_map_sum.rs
/*
 *  Sequential implementation of π by quadrature using imperative approach.
 *
 *  Copyright © 2013–2017, 2019  Russel Winder
 */

use time::precise_time_s;
use output::output;

fn main() {
    let n = 1_000_000_000u64;
    let delta = 1.0 / n as f64;
    let start_time = precise_time_s();
    let pi = 4.0 * delta * (1 .. n).map(|i| {
        let x = (i as f64 - 0.5) * delta;
        1.0 / (1.0 + x * x)
    }).sum::<f64>();
    let elapse_time = precise_time_s() - start_time;
    output("Sequential Map Sum", pi, n, elapse_time)
}
